src/app/InteractionModelEngine.cpp

```diff
@@ -240,9 +254,64 @@ DispatchSingleClusterCommand(chip::ClusterId aClusterId, chip::CommandId aComman
         "Default DispatchSingleClusterCommand is called, this should be replaced by actual dispatched for cluster commands");
 }
 
+CHIP_ERROR __attribute__((weak)) ReadSingleClusterData(AttributePathParams & aAttributePathParams, TLV::TLVWriter & aWriter)
+{
+    ChipLogDetail(DataManagement,
+                  "Received Cluster Command: Cluster=%" PRIx16 " NodeId=%" PRIx64 " Endpoint=%" PRIx8 " FieldId=%" PRIx8
+                  " ListIndex=%" PRIx8,
+                  aAttributePathParams.mClusterId, aAttributePathParams.mNodeId, aAttributePathParams.mEndpointId,
+                  aAttributePathParams.mFieldId, aAttributePathParams.mListIndex);
+    ChipLogError(DataManagement,
+                 "Default ReadSingleClusterData is called, this should be replaced by actual dispatched for cluster");
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR __attribute__((weak)) WriteSingleClusterData(AttributePathParams & aAttributePathParams, TLV::TLVReader & aReader)
+{
+    ChipLogDetail(DataManagement,
+                  "Received Cluster Attribute: Cluster=%" PRIx16 " NodeId=%" PRIx64 " Endpoint=%" PRIx8 " FieldId=%" PRIx8,
+                  " ListIndex=%" PRIx8, aAttributePathParams.mClusterId, aAttributePathParams.mNodeId,
+                  aAttributePathParams.mEndpointId, aAttributePathParams.mFieldId, aAttributePathParams.mListIndex);
+    ChipLogError(DataManagement,
+                 "Default WriteSingleClusterData is called, this should be replaced by actual dispatched for cluster");
+    return CHIP_NO_ERROR;
+}
+
 uint16_t InteractionModelEngine::GetReadClientArrayIndex(const ReadClient * const apReadClient) const
 {
     return static_cast<uint16_t>(apReadClient - mReadClients);
 }
+
+void InteractionModelEngine::ReleaseClusterInfoList(ClusterInfo * apClusterInfo)
+{
+    ClusterInfo * lastClusterInfo = apClusterInfo;
+    if (lastClusterInfo == nullptr)
+    {
+        return;
+    }
+
+    while (lastClusterInfo != nullptr && lastClusterInfo->mpNext != nullptr)
+    {
+        lastClusterInfo->ClearDirty();
+        lastClusterInfo = lastClusterInfo->mpNext;
+    }
+    lastClusterInfo->ClearDirty();
+    lastClusterInfo->mpNext   = mpNextAvaiableClusterInfo;
+    mpNextAvaiableClusterInfo = apClusterInfo;
+}
+
+CHIP_ERROR InteractionModelEngine::PushFront(ClusterInfo *& aClusterInfo, AttributePathParams & aAttributePathParams)
+{
+    if (mpNextAvaiableClusterInfo == nullptr)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    aClusterInfo                       = mpNextAvaiableClusterInfo;
```

Comment:
this seems to lose aClusterInfo.

I would expect that aClusterInfo would get pushed one element in front. Something like:


```
ClusterInfo *tmp = aClusterInfo;
aClusterInfo = mpNextAvailableClusterInfo;
mpNextAvaiableClusterInfo          = mpNextAvaiableClusterInfo->mpNext;
aClusterInfo.mpNext = tmp;
aClusterInfo.mAttributePathParams = aAttributePathParams;
```

could you add unit tests? this code went through several iterations and I am concerned that it could have bugs that are untested

Reply:
done, thanks